source/Irrlicht/CBAWMeshWriter.cpp

```diff
@@ -0,0 +1,269 @@
+					exportAsBlob(reinterpret_cast<core::ICPUBuffer*>(m_headers[i].handle), i, _file);
+					break;
+				case core::Blob::EBT_DATA_FORMAT_DESC:
+					exportAsBlob(reinterpret_cast<IMeshDataFormatDesc<core::ICPUBuffer>*>(m_headers[i].handle), i, _file);
+					break;
+				case core::Blob::EBT_TEXTURE_PATH:
+					exportAsBlob(reinterpret_cast<video::IVirtualTexture*>(m_headers[i].handle), i, _file);
+					break;
+				}
+			}
+
+			const size_t prevPos = _file->getPos();
+
+			// overwrite offsets
+			_file->seek(OFFSETS_FILE_OFFSET);
+			_file->write(m_offsets.const_pointer(), m_offsets.size() * sizeof(m_offsets[0]));
+			// overwrite headers
+			_file->seek(HEADERS_FILE_OFFSET);
+			_file->write(m_headers.const_pointer(), m_headers.size() * sizeof(core::BlobHeader));
+
+			_file->seek(prevPos);
+
+			return true;
+		}
+
+		uint32_t CBAWMeshWriter::genHeaders(ICPUMesh * _mesh)
+		{
+			m_headers.clear();
+			std::set<const IReferenceCounted*> countedObjects;
+			for (uint32_t i = 0; i < _mesh->getMeshBufferCount(); ++i)
+			{
+				const ICPUMeshBuffer * const meshBuffer = _mesh->getMeshBuffer(i);
+				const IMeshDataFormatDesc<core::ICPUBuffer> * const desc = meshBuffer->getMeshDataAndFormat();
+
+				if (!meshBuffer || !desc)
+					continue;
+
+				if (countedObjects.find(meshBuffer) == countedObjects.end())
+				{
+					core::BlobHeader bh;
+					bh.handle = reinterpret_cast<uint64_t>(meshBuffer);
+					bh.compressionType = core::Blob::EBCT_RAW;
+					bh.blobType = core::Blob::EBT_MESH_BUFFER;
+					bh.blobSize = bh.blobSizeDecompr = // size of significant members of ICPUMeshBuffer
+						sizeof(uint64_t) + sizeof(video::E_INDEX_TYPE) + 2 * sizeof(uint32_t) + 2 * sizeof(size_t) + sizeof(uint64_t) + sizeof(E_PRIMITIVE_TYPE) + sizeof(E_VERTEX_ATTRIBUTE_ID)
+						+ sizeof(video::SMaterial)
+						+ sizeof(core::aabbox3df);
+					m_headers.push_back(bh);
+					countedObjects.emplace(meshBuffer);
+
+					const video::SMaterial & mat = meshBuffer->getMaterial();
+					for (int tid = 0; tid < _IRR_MATERIAL_MAX_TEXTURES_; ++tid) // texture path blob headers
+					{
+						video::IVirtualTexture* texture = mat.getTexture(tid);
+						if (mat.getTexture(tid) && countedObjects.find(texture) == countedObjects.end())
+						{
+							bh.handle = reinterpret_cast<uint64_t>(texture);
+							bh.compressionType = core::Blob::EBCT_RAW;
+							bh.blobType = core::Blob::EBT_TEXTURE_PATH;
+							bh.blobSize = bh.blobSizeDecompr = 0; // must be overwritten while exporting blobs
+							m_headers.push_back(bh);
+							countedObjects.emplace(texture);
+						}
+						else continue;
+					}
+				}
+
+				if (countedObjects.find(desc) == countedObjects.end())
+				{
+					core::BlobHeader bh;
+					bh.handle = reinterpret_cast<uint64_t>(desc);
+					bh.compressionType = core::Blob::EBCT_RAW;
+					bh.blobType = core::Blob::EBT_DATA_FORMAT_DESC;
+					bh.blobSize = bh.blobSizeDecompr = // size of members of ICPUMeshDataFormatDesc
```

Comment:
This is complicated to maintain, maybe we should set blobSizeDecompr later?

Or at least as sizeof(IMeshDataFormatDescBlobV1) ?